![alt text](./image_files/pageheader_rose2_babies.jpg)

# Data Science in Medicine using Python

### Author: Dr Gusztav Belteki

## 1. Review of homework

### A more selection

A condensed code of all we did during the last lecture

... This can be difficult to read, particularly when you learn Python and pandas

Use comments!

In [1]:
%%time

import os
import pandas as pd

path = os.path.join('data', 'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip')
data = pd.read_csv(path, parse_dates = [['Date', 'Time']])
data = data.set_index('Date_Time')
new_columns_3 = [item[5:] for item in data.columns if item.startswith('5001')]
new_columns_3 = ['Time [ms]', 'Rel.Time [s]'] + new_columns_3
data.columns = new_columns_3
data = data.resample('1S').mean()
columns_to_drop = ['Tispon [s]', 'I:Espon (I-Part) [no unit]', 'I:Espon (E-Part) [no unit]']
data = data.drop(columns_to_drop, axis = 1)

CPU times: user 3.52 s, sys: 774 ms, total: 4.29 s
Wall time: 4.55 s


In [2]:
data.head()

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2020-11-02 13:42:39  1.604325e+12           0.0          0.2         0.21   
2020-11-02 13:42:40  1.604325e+12           0.5          0.2         0.21   
2020-11-02 13:42:41  1.604325e+12           1.5          0.2         0.21   
2020-11-02 13:42:42  1.604325e+12           2.5          0.2         0.21   
2020-11-02 13:42:43  1.604325e+12           3.5          0.2         0.20   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2020-11-02 13:42:39          0.21         146.0              0.0   
2020-11-02 13:42:40          0.22         146.0              0.0   
2020-11-02 13:42:41          0.26         154.0              0.0   
2020-11-02 13:42:42          0.29         154.0              0.0   
2020-11-02 13:42:43          0.40         193.0              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2020-11-02 13:42:39            128.0              0.2              6.9  ...   
2020-11-02 13:42:40            128.0              0.2              6.9  ...   
2020-11-02 13:42:41            136.0              0.2              6.9  ...   
2020-11-02 13:42:42            136.0              0.2              6.9  ...   
2020-11-02 13:42:43            177.0              0.2              6.9  ...   

                     FiO2 [%]  VTspon [mL]  E [mbar/L]  TC [s]  TCe [s]  \
Date_Time                                                                 
2020-11-02 13:42:39      25.0          NaN         NaN     NaN      NaN   
2020-11-02 13:42:40      25.0          0.0      4774.0    0.03     0.11   
2020-11-02 13:42:41      25.0          0.0      4597.0    0.03     0.21   
2020-11-02 13:42:42      25.0          0.0      3798.0    0.04     0.12   
2020-11-02 13:42:43      25.0          0.0      3408.0    0.05     0.19   

                     C20/Cdyn [no unit]  VTe [mL]  VTi [mL]  EIP [mbar]  \
Date_Time                                                                 
2020-11-02 13:42:39                 NaN       NaN       NaN         NaN   
2020-11-02 13:42:40                0.55       3.4       3.4        21.0   
2020-11-02 13:42:41                0.55       8.2       3.6        22.0   
2020-11-02 13:42:42                0.66       4.1       3.5        19.0   
2020-11-02 13:42:43                0.66       5.5       3.8        18.0   

                     MVleak [L/min]  
Date_Time                            
2020-11-02 13:42:39             NaN  
2020-11-02 13:42:40             0.0  
2020-11-02 13:42:41             0.0  
2020-11-02 13:42:42             0.0  
2020-11-02 13:42:43             0.0  

[5 rows x 42 columns]

Now that the index is time stamps you can select time periodsm

In [3]:
# Select all data during the 1 minute period at 2020-11-03 13:00 

selection = data['2020-11-03 13:00']
selection

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2020-11-03 13:00:00  1.604408e+12  83840.500000         0.26         0.26   
2020-11-03 13:00:01  1.604408e+12  83841.500000         0.26         0.26   
2020-11-03 13:00:02  1.604408e+12  83842.500000         0.26         0.26   
2020-11-03 13:00:03  1.604408e+12  83843.500000         0.26         0.26   
2020-11-03 13:00:04  1.604408e+12  83844.500000         0.26         0.26   
2020-11-03 13:00:05  1.604408e+12  83845.500000         0.25         0.26   
2020-11-03 13:00:06  1.604408e+12  83846.500000         0.25         0.26   
2020-11-03 13:00:07  1.604408e+12  83847.500000         0.25         0.26   
2020-11-03 13:00:08  1.604408e+12  83848.500000         0.25         0.26   
2020-11-03 13:00:09  1.604408e+12  83849.666667         0.25         0.26   
2020-11-03 13:00:10  1.604408e+12  83851.000000         0.25         0.26   
2020-11-03 13:00:11  1.604408e+12  83851.500000         0.25         0.26   
2020-11-03 13:00:12  1.604408e+12  83852.500000         0.25         0.26   
2020-11-03 13:00:13  1.604408e+12  83853.500000         0.25         0.26   
2020-11-03 13:00:14  1.604408e+12  83854.500000         0.25         0.26   
2020-11-03 13:00:15  1.604408e+12  83855.666667         0.25         0.27   
2020-11-03 13:00:16  1.604408e+12  83857.000000         0.25         0.27   
2020-11-03 13:00:17  1.604408e+12  83857.500000         0.25         0.27   
2020-11-03 13:00:18  1.604408e+12  83858.500000         0.25         0.27   
2020-11-03 13:00:19  1.604408e+12  83859.500000         0.25         0.27   
2020-11-03 13:00:20  1.604408e+12  83860.500000         0.25         0.28   
2020-11-03 13:00:21  1.604408e+12  83861.500000         0.25         0.28   
2020-11-03 13:00:22  1.604408e+12  83862.500000         0.25         0.28   
2020-11-03 13:00:23  1.604408e+12  83863.500000         0.25         0.28   
2020-11-03 13:00:24  1.604408e+12  83864.500000         0.25         0.28   
2020-11-03 13:00:25  1.604408e+12  83865.500000         0.25         0.27   
2020-11-03 13:00:26  1.604408e+12  83866.500000         0.25         0.27   
2020-11-03 13:00:27  1.604408e+12  83867.500000         0.25         0.27   
2020-11-03 13:00:28  1.604408e+12  83868.500000         0.25         0.27   
2020-11-03 13:00:29  1.604408e+12  83869.500000         0.25         0.27   
2020-11-03 13:00:30  1.604408e+12  83870.500000         0.25         0.27   
2020-11-03 13:00:31  1.604408e+12  83871.666667         0.25         0.27   
2020-11-03 13:00:32  1.604408e+12  83873.000000         0.25         0.27   
2020-11-03 13:00:33  1.604408e+12  83873.500000         0.25         0.27   
2020-11-03 13:00:34  1.604408e+12  83874.500000         0.26         0.27   
2020-11-03 13:00:35  1.604408e+12  83875.500000         0.26         0.27   
2020-11-03 13:00:36  1.604408e+12  83876.500000         0.26         0.27   
2020-11-03 13:00:37  1.604408e+12  83877.500000         0.27         0.27   
2020-11-03 13:00:38  1.604408e+12  83878.500000         0.27         0.27   
2020-11-03 13:00:39  1.604408e+12  83879.500000         0.27         0.27   
2020-11-03 13:00:40  1.604408e+12  83880.500000         0.28         0.28   
2020-11-03 13:00:41  1.604408e+12  83881.500000         0.28         0.28   
2020-11-03 13:00:42  1.604408e+12  83882.500000         0.28         0.28   
2020-11-03 13:00:43  1.604408e+12  83883.500000         0.28         0.28   
2020-11-03 13:00:44  1.604408e+12  83884.500000         0.28         0.29   
2020-11-03 13:00:45  1.604408e+12  83885.500000         0.28         0.29   
2020-11-03 13:00:46  1.604408e+12  83886.500000         0.27         0.29   
2020-11-03 13:00:47  1.604408e+12  83887.500000         0.27         0.28   
2020-11-03 13:00:48  1.604408e+12  83888.500000         0.27         0.28   
2020-11-03 13:00:49  1.604408e+12  83889.500000         0.27         0.28   
2020-11-03 13:00:50 

In [4]:
len(selection)

60

In [5]:
# Select all data between 2020-11-03 13:00 and and 15:00

selection = data.loc['2020-11-06 13:00' : '2020-11-06 15:00']
selection

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2020-11-06 13:00:00  1.604668e+12      343040.5         0.20         0.21   
2020-11-06 13:00:01  1.604668e+12      343041.5         0.20         0.21   
2020-11-06 13:00:02  1.604668e+12      343042.5         0.20         0.21   
2020-11-06 13:00:03  1.604668e+12      343043.5         0.20         0.21   
2020-11-06 13:00:04  1.604668e+12      343044.5         0.20         0.21   
...                           ...           ...          ...          ...   
2020-11-06 13:27:11  1.604669e+12      344671.5         0.22         0.21   
2020-11-06 13:27:12  1.604669e+12      344672.5         0.23         0.22   
2020-11-06 13:27:13  1.604669e+12      344673.5         0.22         0.22   
2020-11-06 13:27:14  1.604669e+12      344674.5         0.22         0.22   
2020-11-06 13:27:15  1.604669e+12      344675.5         0.22         0.22   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2020-11-06 13:00:00          0.23         171.0              0.0   
2020-11-06 13:00:01          0.24         171.0              0.0   
2020-11-06 13:00:02          0.22         191.0              0.0   
2020-11-06 13:00:03          0.24         191.0              0.0   
2020-11-06 13:00:04          0.24         191.0              0.0   
...                           ...           ...              ...   
2020-11-06 13:27:11          0.22         211.0              0.0   
2020-11-06 13:27:12          0.26         164.0              0.0   
2020-11-06 13:27:13          0.21         164.0              0.0   
2020-11-06 13:27:14          0.21         164.0              0.0   
2020-11-06 13:27:15          0.05         164.0              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2020-11-06 13:00:00            153.0             0.20              6.8  ...   
2020-11-06 13:00:01            153.0             0.20              6.8  ...   
2020-11-06 13:00:02            175.0             0.20              6.8  ...   
2020-11-06 13:00:03            175.0             0.20              6.8  ...   
2020-11-06 13:00:04            175.0             0.20              6.8  ...   
...                              ...              ...              ...  ...   
2020-11-06 13:27:11            193.0             0.22              6.8  ...   
2020-11-06 13:27:12            150.0             0.22              6.9  ...   
2020-11-06 13:27:13            150.0             0.22              6.9  ...   
2020-11-06 13:27:14            150.0             0.22              6.9  ...   
2020-11-06 13:27:15            150.0             0.22              6.9  ...   

                     FiO2 [%]  VTspon [mL]  E [mbar/L]  TC [s]  TCe [s]  \
Date_Time                                                                 
2020-11-06 13:00:00      26.0          0.0      4125.0    0.04     0.12   
2020-11-06 13:00:01      26.0          0.0      4326.0    0.04     0.12   
2020-11-06 13:00:02      26.0          0.0      4248.0    0.04     0.12   
2020-11-06 13:00:03      26.0          0.0      4448.0    0.04     0.12   
2020-11-06 13:00:04      26.0          0.0      4226.0    0.05     0.12   
...                       ...          ...         ...     ...      ...   
2020-11-06 13:27:11      26.0          0.0      4502.0    0.07     0.12   
2020-11-06 13:27:12      26.0          0.0      3879.0    0.05     0.16   
2020-11-06 13:27:13      26.0          0.0      4695.0    0.05     0.21   
2020-11-06 13:27:14      26.0          0.0      4726.0    0.04     0.11   
2020-11-06 13:27:15      26.0          0.0         NaN    0.01     0.15   

                     C20/Cdyn [no unit]  VTe [mL]  VTi [mL]  EIP [mbar]  \
Date_Time                                                 

In [6]:
len(selection)

1636

The recording does not last until 15:00, therefore the selection is actually shorter than 2 hours

In [7]:
# Select all data between 2020-11-03 9:00 and and 11:00

selection = data.loc['2020-11-06 09:00' : '2020-11-06 11:00']
selection

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2020-11-06 09:00:00  1.604653e+12      328640.5         0.22         0.25   
2020-11-06 09:00:01  1.604653e+12      328641.5         0.22         0.25   
2020-11-06 09:00:02  1.604653e+12      328642.5         0.22         0.25   
2020-11-06 09:00:03  1.604653e+12      328643.5         0.22         0.25   
2020-11-06 09:00:04  1.604653e+12      328644.5         0.22         0.25   
...                           ...           ...          ...          ...   
2020-11-06 11:00:55  1.604660e+12      335895.5         0.21         0.20   
2020-11-06 11:00:56  1.604660e+12      335896.5         0.21         0.21   
2020-11-06 11:00:57  1.604660e+12      335897.5         0.21         0.21   
2020-11-06 11:00:58  1.604660e+12      335898.5         0.21         0.21   
2020-11-06 11:00:59  1.604660e+12      335899.5         0.21         0.21   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2020-11-06 09:00:00          0.23         207.0              0.0   
2020-11-06 09:00:01          0.23         207.0              0.0   
2020-11-06 09:00:02          0.23         207.0              0.0   
2020-11-06 09:00:03          0.24         204.0              0.0   
2020-11-06 09:00:04          0.24         204.0              0.0   
...                           ...           ...              ...   
2020-11-06 11:00:55          0.25         133.0              0.0   
2020-11-06 11:00:56          0.25         133.0              0.0   
2020-11-06 11:00:57          0.25         145.0              0.0   
2020-11-06 11:00:58          0.24         145.0              0.0   
2020-11-06 11:00:59          0.23         145.0              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2020-11-06 09:00:00            189.0             0.22              6.9  ...   
2020-11-06 09:00:01            189.0             0.22              6.9  ...   
2020-11-06 09:00:02            189.0             0.22              6.9  ...   
2020-11-06 09:00:03            186.0             0.22              7.0  ...   
2020-11-06 09:00:04            186.0             0.22              7.0  ...   
...                              ...              ...              ...  ...   
2020-11-06 11:00:55            113.0             0.21              6.8  ...   
2020-11-06 11:00:56            113.0             0.21              6.8  ...   
2020-11-06 11:00:57            127.0             0.21              6.8  ...   
2020-11-06 11:00:58            127.0             0.21              6.8  ...   
2020-11-06 11:00:59            127.0             0.21              6.8  ...   

                     FiO2 [%]  VTspon [mL]  E [mbar/L]  TC [s]  TCe [s]  \
Date_Time                                                                 
2020-11-06 09:00:00      26.0          0.0      4361.0    0.05     0.13   
2020-11-06 09:00:01      26.0          0.0      4367.0    0.05     0.13   
2020-11-06 09:00:02      26.0          0.0      4390.0    0.05     0.13   
2020-11-06 09:00:03      26.0          0.0      4296.0    0.05     0.13   
2020-11-06 09:00:04      26.0          0.0      4232.0    0.05     0.13   
...                       ...          ...         ...     ...      ...   
2020-11-06 11:00:55      24.0          0.0      3989.0    0.03     0.11   
2020-11-06 11:00:56      24.0          0.0      3986.0    0.03     0.12   
2020-11-06 11:00:57      24.0          0.0      4039.0    0.03     0.11   
2020-11-06 11:00:58      24.0          0.0      4106.0    0.04     0.12   
2020-11-06 11:00:59      24.0          0.0      4256.0    0.03     0.12   

                     C20/Cdyn [no unit]  VTe [mL]  VTi [mL]  EIP [mbar]  \
Date_Time                                                 

In [8]:
len(selection)

7260

This is 7260 data points because the last minute is also included
This is different from the logic of Python

In [9]:
# Select all data between 2020-11-03 9:00 and and 11:00

selection = data.loc['2020-11-06 09:00' : '2020-11-06 10:59']
selection

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2020-11-06 09:00:00  1.604653e+12      328640.5         0.22         0.25   
2020-11-06 09:00:01  1.604653e+12      328641.5         0.22         0.25   
2020-11-06 09:00:02  1.604653e+12      328642.5         0.22         0.25   
2020-11-06 09:00:03  1.604653e+12      328643.5         0.22         0.25   
2020-11-06 09:00:04  1.604653e+12      328644.5         0.22         0.25   
...                           ...           ...          ...          ...   
2020-11-06 10:59:55  1.604660e+12      335835.5         0.20         0.21   
2020-11-06 10:59:56  1.604660e+12      335836.5         0.20         0.21   
2020-11-06 10:59:57  1.604660e+12      335837.5         0.20         0.21   
2020-11-06 10:59:58  1.604660e+12      335838.5         0.20         0.21   
2020-11-06 10:59:59  1.604660e+12      335839.5         0.20         0.21   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2020-11-06 09:00:00          0.23         207.0              0.0   
2020-11-06 09:00:01          0.23         207.0              0.0   
2020-11-06 09:00:02          0.23         207.0              0.0   
2020-11-06 09:00:03          0.24         204.0              0.0   
2020-11-06 09:00:04          0.24         204.0              0.0   
...                           ...           ...              ...   
2020-11-06 10:59:55          0.25         154.0              0.0   
2020-11-06 10:59:56          0.24         154.0              0.0   
2020-11-06 10:59:57          0.24         151.0              0.0   
2020-11-06 10:59:58          0.24         151.0              0.0   
2020-11-06 10:59:59          0.26         127.0              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2020-11-06 09:00:00            189.0             0.22              6.9  ...   
2020-11-06 09:00:01            189.0             0.22              6.9  ...   
2020-11-06 09:00:02            189.0             0.22              6.9  ...   
2020-11-06 09:00:03            186.0             0.22              7.0  ...   
2020-11-06 09:00:04            186.0             0.22              7.0  ...   
...                              ...              ...              ...  ...   
2020-11-06 10:59:55            136.0             0.20              6.8  ...   
2020-11-06 10:59:56            136.0             0.20              6.8  ...   
2020-11-06 10:59:57            134.0             0.20              6.8  ...   
2020-11-06 10:59:58            134.0             0.20              6.8  ...   
2020-11-06 10:59:59            108.0             0.20              6.8  ...   

                     FiO2 [%]  VTspon [mL]  E [mbar/L]  TC [s]  TCe [s]  \
Date_Time                                                                 
2020-11-06 09:00:00      26.0          0.0      4361.0    0.05     0.13   
2020-11-06 09:00:01      26.0          0.0      4367.0    0.05     0.13   
2020-11-06 09:00:02      26.0          0.0      4390.0    0.05     0.13   
2020-11-06 09:00:03      26.0          0.0      4296.0    0.05     0.13   
2020-11-06 09:00:04      26.0          0.0      4232.0    0.05     0.13   
...                       ...          ...         ...     ...      ...   
2020-11-06 10:59:55      24.0          0.0      4031.0    0.04     0.12   
2020-11-06 10:59:56      24.0          0.0      4170.0    0.04     0.12   
2020-11-06 10:59:57      24.0          0.0      4095.0    0.04     0.12   
2020-11-06 10:59:58      24.0          0.0      4193.0    0.04     0.12   
2020-11-06 10:59:59      24.0          0.0      3875.0    0.04     0.11   

                     C20/Cdyn [no unit]  VTe [mL]  VTi [mL]  EIP [mbar]  \
Date_Time                                                 

In [10]:
len(selection)

7200

This will also work if there are missing data, therefore it is preferable to choosing simply 7200 rows

In [11]:
# Select all data between 2020-11-03 13:00 and and 15:00 and limit it to 
# "MVe [L/min]" and "MVi [L/min]" columns only

selection = data['2020-11-06 13:00' : '2020-11-06 15:00'][['MVe [L/min]', 'MVi [L/min]']]
selection

MVe [L/min]  MVi [L/min]
Date_Time                                    
2020-11-06 13:00:00         0.20         0.21
2020-11-06 13:00:01         0.20         0.21
2020-11-06 13:00:02         0.20         0.21
2020-11-06 13:00:03         0.20         0.21
2020-11-06 13:00:04         0.20         0.21
...                          ...          ...
2020-11-06 13:27:11         0.22         0.21
2020-11-06 13:27:12         0.23         0.22
2020-11-06 13:27:13         0.22         0.22
2020-11-06 13:27:14         0.22         0.22
2020-11-06 13:27:15         0.22         0.22

[1636 rows x 2 columns]

### B. Import and process the `Radiometer_TCM_sample_data.xlsx'` file containing transcutaneous CO2 data into a Dataframe named "data_2"

The file is located in the "data" subfolder

*Hint: the first 8 rows contain text rather than data and need to be skipped*

In [12]:
import pandas as pd
import os

path = os.path.join('data', 'Radiometer_TCM_sample_data.xlsx')
data_2 = pd.read_excel(path, skiprows = 8)

In [13]:
data_2.head()

Time                  Event O2 kPa CO2 kPa Power mW  Temp ∞C  \
0 2016-05-25 11:40:12          Session start      -       -        -     42.5   
1 2016-05-25 11:40:12                    NaN      -       -        -     42.5   
2 2016-05-25 11:40:12  Arterialization start      -       -        -     42.5   
3 2016-05-25 11:40:13                    NaN      -       3      304     42.5   
4 2016-05-25 11:40:14                    NaN      -       3      308     42.5   

  SpO2 % Pulse bmp  
0      -         -  
1      -         -  
2      -         -  
3      0         0  
4      0         0

Check the data types of the various columns

In [14]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1036 entries, 0 to 1035
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Time       1036 non-null   datetime64[ns]
 1   Event      6 non-null      object        
 2   O2 kPa     1036 non-null   object        
 3   CO2 kPa    1036 non-null   object        
 4   Power mW   1036 non-null   object        
 5   Temp ∞C    1036 non-null   float64       
 6   SpO2 %     1036 non-null   object        
 7   Pulse bmp  1036 non-null   object        
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 64.9+ KB


Now the 'Time` column seems to be in the correct format. Set it as index

In [15]:
data_2 = data_2.set_index('Time')
data_2.head()

Event O2 kPa CO2 kPa Power mW  Temp ∞C  \
Time                                                                          
2016-05-25 11:40:12          Session start      -       -        -     42.5   
2016-05-25 11:40:12                    NaN      -       -        -     42.5   
2016-05-25 11:40:12  Arterialization start      -       -        -     42.5   
2016-05-25 11:40:13                    NaN      -       3      304     42.5   
2016-05-25 11:40:14                    NaN      -       3      308     42.5   

                    SpO2 % Pulse bmp  
Time                                  
2016-05-25 11:40:12      -         -  
2016-05-25 11:40:12      -         -  
2016-05-25 11:40:12      -         -  
2016-05-25 11:40:13      0         0  
2016-05-25 11:40:14      0         0

 ...but most other columns are just objects. First check what values do the columns contain

In [16]:
data_2.columns

Index(['Event', 'O2 kPa', 'CO2 kPa', 'Power mW', 'Temp ∞C', 'SpO2 %',
       'Pulse bmp'],
      dtype='object')

In [17]:
for column in data_2.columns:
    print(column,data_2[column].unique(), '\n')

Event ['Session start' nan 'Arterialization start' 'Patient Details Changed'
 'Arterialization end' 'Calibrating' 'Session end'] 

O2 kPa ['-'] 

CO2 kPa ['-' 3 2.9 2.8 2.7 2.6 2.5 2.4 2.3 2.2 2.1 2 1.9 1.8 1.7 1.6 1.5 1.4 3.1
 3.2 3.3 3.4 3.5 3.6 3.7 3.8 3.9 4 4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9 5
 5.1 5.2 5.3 1.3 1.2 1.1 1 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1 0] 

Power mW ['-' 304 308 857 648 480 427 374 393 548 675 831 797 702 544 454 735 784
 710 657 676 607 602 575 569 559 567 549 541 542 524 523 518 507 490 510
 515 499 509 494 491 483 486 482 475 477 476 468 473 466 458 444 450 447
 443 438 448 436 430 445 446 421 435 432 418 431 423 424 417 411 409 408
 414 405 394 420 412 406 384 402 399 403 392 395 380 388 386 390 389 371
 381 387 378 375 363 372 370 365 367 366 368 369 349 362 361 351 353 358
 348 355 334 345 342 339 352 350 341 347 326 338 340 335 332 346 356 344
 337 343 336 333 328 331 330 322 318 313 317 309 319 314 312 307 301 310
 311 306 303 302 298 291 300 299 297 29

... to be continued

## 2. What if we want to work on multiple files ?

Repeat what was done last time on a single file

### What is the problem with these data

- Indexed by numbers only (uninformative, it should be indexed by date and time)
- Date and time are in separate columnns
- Date and time formats are not appropriate
- Column names are too long and difficult to read
- Lots of `na` values 
- half of every row is empty
- Some columns have barely any informative values
- Some values are not meaningful (e.g. tidal volume should be mL/kg not mL)

*We will deal with all these issues*

#### Importing a list of files

In [18]:
# Remember: lists can be a collection of anything

lst = ['a', 'b', 'c', 'd']
lst

['a', 'b', 'c', 'd']

In [19]:
# You can go through a list using a for loop

for item in lst: # `item` can be replaced by any word; note the colon at the end
    print(item) # note the indentation

a
b
c
d


In [20]:
lst = []
lst.append('a') # adding elements to list
lst.append('b')
lst

['a', 'b']

##### Now do this for a more complicated stuff

In [21]:
%%time

import os
import pandas as pd

filelist = ['CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip', 
            'CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip',
            'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip',]

data_list = [] # Creates an empty list

for file in filelist:
    print(file) # This will let you know what is happening
    path = os.path.join('data', file,)
    data_list.append(pd.read_csv(path))

CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 4.84 s, sys: 650 ms, total: 5.49 s
Wall time: 5.6 s


In [22]:
len(data_list)

3

In [23]:
data_list[0]

Time [ms]        Date          Time  Rel.Time [s]  \
0       1547469721191  2019-01-14  12:42:01.191             0   
1       1547469721332  2019-01-14  12:42:01.332             0   
2       1547469722191  2019-01-14  12:42:02.191             1   
3       1547469722316  2019-01-14  12:42:02.316             1   
4       1547469723191  2019-01-14  12:42:03.191             2   
...               ...         ...           ...           ...   
209895  1547574652322  2019-01-15  17:50:52.322        104931   
209896  1547574653197  2019-01-15  17:50:53.197        104932   
209897  1547574653306  2019-01-15  17:50:53.306        104932   
209898  1547574654213  2019-01-15  17:50:54.213        104933   
209899  1547574654322  2019-01-15  17:50:54.322        104933   

        5001|MVe [L/min]  5001|MVi [L/min]  5001|Cdyn [L/bar]  \
0                   0.15              0.13               0.29   
1                    NaN               NaN                NaN   
2                   0.15              0.13               0.29   
3                    NaN               NaN                NaN   
4                   0.15              0.13               0.29   
...                  ...               ...                ...   
209895               NaN               NaN                NaN   
209896              0.00              2.38                NaN   
209897               NaN               NaN                NaN   
209898              0.00              2.38                NaN   
209899               NaN               NaN                NaN   

        5001|R [mbar/L/s]  5001|MVespon [L/min]  5001|Rpat [mbar/L/s]  ...  \
0                   142.0                   0.0                 129.0  ...   
1                     NaN                   NaN                   NaN  ...   
2                   152.0                   0.0                 139.0  ...   
3                     NaN                   NaN                   NaN  ...   
4                   152.0                   0.0                 139.0  ...   
...                   ...                   ...                   ...  ...   
209895                NaN                   NaN                   NaN  ...   
209896                NaN                   0.0                   NaN  ...   
209897                NaN                   NaN                   NaN  ...   
209898                NaN                   0.0                   NaN  ...   
209899                NaN                   NaN                   NaN  ...   

        5001|TCe [s]  5001|C20/Cdyn [no unit]  5001|VTe [mL]  5001|VTi [mL]  \
0                NaN                      NaN            NaN            NaN   
1               0.11                     0.58            2.7            2.4   
2                NaN                      NaN            NaN            NaN   
3               0.11                     0.66            2.8            2.6   
4                NaN                      NaN            NaN            NaN   
...              ...                      ...            ...            ...   
209895           NaN                      NaN            NaN            NaN   
209896           NaN                      NaN            NaN            NaN   
209897           NaN                      NaN            NaN            NaN   
209898           NaN                      NaN            NaN            NaN   
209899           NaN                      NaN            NaN            NaN   

        5001|EIP [mbar]  5001|MVleak [L/min]  5001|RRspon [1/min]  \
0                   NaN                  NaN                  NaN   
1                  16.0                  0.0                  NaN   
2                   NaN                  NaN                  NaN   
3                  16.0                  0.0                  NaN   
4                   NaN                  NaN                  NaN   
...                 ...                  ...                  ...   
209895              NaN                  0.0                  NaN   
209896      

In [24]:
data_list[0].head()

Time [ms]        Date          Time  Rel.Time [s]  5001|MVe [L/min]  \
0  1547469721191  2019-01-14  12:42:01.191             0              0.15   
1  1547469721332  2019-01-14  12:42:01.332             0               NaN   
2  1547469722191  2019-01-14  12:42:02.191             1              0.15   
3  1547469722316  2019-01-14  12:42:02.316             1               NaN   
4  1547469723191  2019-01-14  12:42:03.191             2              0.15   

   5001|MVi [L/min]  5001|Cdyn [L/bar]  5001|R [mbar/L/s]  \
0              0.13               0.29              142.0   
1               NaN                NaN                NaN   
2              0.13               0.29              152.0   
3               NaN                NaN                NaN   
4              0.13               0.29              152.0   

   5001|MVespon [L/min]  5001|Rpat [mbar/L/s]  ...  5001|TCe [s]  \
0                   0.0                 129.0  ...           NaN   
1                   NaN                   NaN  ...          0.11   
2                   0.0                 139.0  ...           NaN   
3                   NaN                   NaN  ...          0.11   
4                   0.0                 139.0  ...           NaN   

   5001|C20/Cdyn [no unit]  5001|VTe [mL]  5001|VTi [mL]  5001|EIP [mbar]  \
0                      NaN            NaN            NaN              NaN   
1                     0.58            2.7            2.4             16.0   
2                      NaN            NaN            NaN              NaN   
3                     0.66            2.8            2.6             16.0   
4                      NaN            NaN            NaN              NaN   

   5001|MVleak [L/min]  5001|RRspon [1/min]  5001|Tispon [s]  \
0                  NaN                  NaN              NaN   
1                  0.0                  NaN              NaN   
2                  NaN                  NaN              NaN   
3                  0.0                  NaN              NaN   
4                  NaN                  NaN              NaN   

   5001|I:Espon (I-Part) [no unit]  5001|I:Espon (E-Part) [no unit]  
0                              NaN                              NaN  
1                              NaN                              NaN  
2                              NaN                              NaN  
3                              NaN                              NaN  
4                              NaN                              NaN  

[5 rows x 47 columns]

In [25]:
data_list[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209900 entries, 0 to 209899
Data columns (total 47 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   Time [ms]                        209900 non-null  int64  
 1   Date                             209900 non-null  object 
 2   Time                             209900 non-null  object 
 3   Rel.Time [s]                     209900 non-null  int64  
 4   5001|MVe [L/min]                 104893 non-null  float64
 5   5001|MVi [L/min]                 104899 non-null  float64
 6   5001|Cdyn [L/bar]                102860 non-null  float64
 7   5001|R [mbar/L/s]                103751 non-null  float64
 8   5001|MVespon [L/min]             104893 non-null  float64
 9   5001|Rpat [mbar/L/s]             102230 non-null  float64
 10  5001|MVemand [L/min]             104893 non-null  float64
 11  5001|FlowDev [L/min]             104899 non-null  float64
 12  50

##### It is a good idea to report the duration the various steps take

In [26]:
%%time

import datetime # with this you can report the time your program's execution
import os
import pandas as pd


filelist = ['CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip', 
            'CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip',
            'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip',]

data_list = [] # Creates an empty list

for file in filelist:
    print(datetime.datetime.now(), file)
    path = os.path.join('data', file,)
    data_list.append(pd.read_csv(path))

2021-04-20 09:12:04.900306 CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
2021-04-20 09:12:05.930402 CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
2021-04-20 09:12:07.046032 CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 4.44 s, sys: 646 ms, total: 5.08 s
Wall time: 5.15 s


The problem is that we have to produce the list of file names manually. This is error prone and can be a lot of work if there are many files. There must be a better way...

In [27]:
flist = os.listdir('data') # we could also use os.listdir(os.path.join('data'))
flist

['CsvLogBase_2020-11-06_132719.623_slow_Setting.csv',
 'Radiometer_TCM_sample_data.xlsx',
 'CsvLogBase_2019-01-14_124200.144_slow_AlarmState.csv',
 'CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip',
 '.DS_Store',
 'CsvLogBase_2020-11-06_132719.623_slow_AlarmSetting.csv',
 'CsvLogBase_2019-01-14_124200.144_slow_Setting.csv',
 'Info.txt',
 'karamazov_brothers.txt',
 'CsvLogBase_2019-01-16_090910.423_slow_AlarmState.csv',
 'CsvLogBase_2019-01-16_090910.423_slow_AlarmSetting.csv',
 'CsvLogBase_2020-11-06_132719.623_slow_AlarmState.csv',
 'CsvLogBase_2019-01-14_124200.144_slow_AlarmSetting.csv',
 'README.md',
 'CsvLogBase_2020-11-02_134238.904_slow_Text.csv',
 'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip',
 'CsvLogBase_2020-11-02_134238.904_slow_AlarmState.csv',
 'CsvLogBase_2019-01-14_124200.144_slow_Text.csv',
 'CsvLogBase_2020-11-02_134238.904_slow_AlarmSetting.csv',
 'CsvLogBase_2019-01-16_090910.423_slow_Setting.csv',
 'CsvLogBase_2019-01-16_090910.423_slow_Text

In [28]:
flist = [fle for fle in os.listdir('data') if 'slow_Measurement' in fle]
flist

['CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip',
 'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip',
 'CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip']

In [29]:
%%time

import datetime # with this you can report the time your program's execution
import os
import pandas as pd

flist = [fle for fle in os.listdir('data') if 'slow_Measurement' in fle]

data_list = [] # Creates an empty list

for file in filelist:
    print(datetime.datetime.now(), file)
    path = os.path.join('data', file,)
    data_list.append(pd.read_csv(path))

2021-04-20 09:12:10.080939 CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
2021-04-20 09:12:10.924540 CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
2021-04-20 09:12:11.726337 CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 3.81 s, sys: 576 ms, total: 4.39 s
Wall time: 4.39 s


It is also a problem that we can only identify the individual elements by position.
It would be better to identify them by name.
This can be done with dictionaries.

In [30]:
%%time

import datetime
import os
import pandas as pd

flist = [fle for fle in os.listdir('data') if 'slow_Measurement' in fle]

data_dict = {} # Creates an empty dictionary

for file in filelist:
    print(datetime.datetime.now(), file)
    path = os.path.join('data', file,)
    data_dict[file] = pd.read_csv(path)

2021-04-20 09:12:14.403989 CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
2021-04-20 09:12:15.265543 CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
2021-04-20 09:12:16.076067 CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 3.96 s, sys: 551 ms, total: 4.52 s
Wall time: 4.56 s


In [31]:
len(data_dict)

3

In [32]:
data_dict['CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip']

Time [ms]        Date          Time  Rel.Time [s]  \
0       1547469721191  2019-01-14  12:42:01.191             0   
1       1547469721332  2019-01-14  12:42:01.332             0   
2       1547469722191  2019-01-14  12:42:02.191             1   
3       1547469722316  2019-01-14  12:42:02.316             1   
4       1547469723191  2019-01-14  12:42:03.191             2   
...               ...         ...           ...           ...   
209895  1547574652322  2019-01-15  17:50:52.322        104931   
209896  1547574653197  2019-01-15  17:50:53.197        104932   
209897  1547574653306  2019-01-15  17:50:53.306        104932   
209898  1547574654213  2019-01-15  17:50:54.213        104933   
209899  1547574654322  2019-01-15  17:50:54.322        104933   

        5001|MVe [L/min]  5001|MVi [L/min]  5001|Cdyn [L/bar]  \
0                   0.15              0.13               0.29   
1                    NaN               NaN                NaN   
2                   0.15              0.13               0.29   
3                    NaN               NaN                NaN   
4                   0.15              0.13               0.29   
...                  ...               ...                ...   
209895               NaN               NaN                NaN   
209896              0.00              2.38                NaN   
209897               NaN               NaN                NaN   
209898              0.00              2.38                NaN   
209899               NaN               NaN                NaN   

        5001|R [mbar/L/s]  5001|MVespon [L/min]  5001|Rpat [mbar/L/s]  ...  \
0                   142.0                   0.0                 129.0  ...   
1                     NaN                   NaN                   NaN  ...   
2                   152.0                   0.0                 139.0  ...   
3                     NaN                   NaN                   NaN  ...   
4                   152.0                   0.0                 139.0  ...   
...                   ...                   ...                   ...  ...   
209895                NaN                   NaN                   NaN  ...   
209896                NaN                   0.0                   NaN  ...   
209897                NaN                   NaN                   NaN  ...   
209898                NaN                   0.0                   NaN  ...   
209899                NaN                   NaN                   NaN  ...   

        5001|TCe [s]  5001|C20/Cdyn [no unit]  5001|VTe [mL]  5001|VTi [mL]  \
0                NaN                      NaN            NaN            NaN   
1               0.11                     0.58            2.7            2.4   
2                NaN                      NaN            NaN            NaN   
3               0.11                     0.66            2.8            2.6   
4                NaN                      NaN            NaN            NaN   
...              ...                      ...            ...            ...   
209895           NaN                      NaN            NaN            NaN   
209896           NaN                      NaN            NaN            NaN   
209897           NaN                      NaN            NaN            NaN   
209898           NaN                      NaN            NaN            NaN   
209899           NaN                      NaN            NaN            NaN   

        5001|EIP [mbar]  5001|MVleak [L/min]  5001|RRspon [1/min]  \
0                   NaN                  NaN                  NaN   
1                  16.0                  0.0                  NaN   
2                   NaN                  NaN                  NaN   
3                  16.0                  0.0                  NaN   
4                   NaN                  NaN                  NaN   
...                 ...                  ...                  ...   
209895              NaN                  0.0                  NaN   
209896      

In [33]:
data_dict.keys()

dict_keys(['CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip', 'CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip', 'CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip'])

These indices are too long, use shorter but still specific indices

In [34]:
%%time

import datetime
import os
import pandas as pd

flist = [fle for fle in os.listdir('data') if 'slow_Measurement' in fle]

data_dict = {} # Creates an empty dictionary

for file in filelist:
    print(datetime.datetime.now(), file)
    path = os.path.join('data', file,)
    tag = file[11:-25] # Create a shorter tag
    data_dict[tag] = pd.read_csv(path)

2021-04-20 09:12:19.102623 CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
2021-04-20 09:12:20.119669 CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
2021-04-20 09:12:20.956488 CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 4.01 s, sys: 582 ms, total: 4.59 s
Wall time: 4.64 s


In [35]:
data_dict.keys()

dict_keys(['2019-01-14_124200.144', '2019-01-16_090910.423', '2020-11-02_134238.904'])

In [36]:
data_dict['2019-01-14_124200.144']

Time [ms]        Date          Time  Rel.Time [s]  \
0       1547469721191  2019-01-14  12:42:01.191             0   
1       1547469721332  2019-01-14  12:42:01.332             0   
2       1547469722191  2019-01-14  12:42:02.191             1   
3       1547469722316  2019-01-14  12:42:02.316             1   
4       1547469723191  2019-01-14  12:42:03.191             2   
...               ...         ...           ...           ...   
209895  1547574652322  2019-01-15  17:50:52.322        104931   
209896  1547574653197  2019-01-15  17:50:53.197        104932   
209897  1547574653306  2019-01-15  17:50:53.306        104932   
209898  1547574654213  2019-01-15  17:50:54.213        104933   
209899  1547574654322  2019-01-15  17:50:54.322        104933   

        5001|MVe [L/min]  5001|MVi [L/min]  5001|Cdyn [L/bar]  \
0                   0.15              0.13               0.29   
1                    NaN               NaN                NaN   
2                   0.15              0.13               0.29   
3                    NaN               NaN                NaN   
4                   0.15              0.13               0.29   
...                  ...               ...                ...   
209895               NaN               NaN                NaN   
209896              0.00              2.38                NaN   
209897               NaN               NaN                NaN   
209898              0.00              2.38                NaN   
209899               NaN               NaN                NaN   

        5001|R [mbar/L/s]  5001|MVespon [L/min]  5001|Rpat [mbar/L/s]  ...  \
0                   142.0                   0.0                 129.0  ...   
1                     NaN                   NaN                   NaN  ...   
2                   152.0                   0.0                 139.0  ...   
3                     NaN                   NaN                   NaN  ...   
4                   152.0                   0.0                 139.0  ...   
...                   ...                   ...                   ...  ...   
209895                NaN                   NaN                   NaN  ...   
209896                NaN                   0.0                   NaN  ...   
209897                NaN                   NaN                   NaN  ...   
209898                NaN                   0.0                   NaN  ...   
209899                NaN                   NaN                   NaN  ...   

        5001|TCe [s]  5001|C20/Cdyn [no unit]  5001|VTe [mL]  5001|VTi [mL]  \
0                NaN                      NaN            NaN            NaN   
1               0.11                     0.58            2.7            2.4   
2                NaN                      NaN            NaN            NaN   
3               0.11                     0.66            2.8            2.6   
4                NaN                      NaN            NaN            NaN   
...              ...                      ...            ...            ...   
209895           NaN                      NaN            NaN            NaN   
209896           NaN                      NaN            NaN            NaN   
209897           NaN                      NaN            NaN            NaN   
209898           NaN                      NaN            NaN            NaN   
209899           NaN                      NaN            NaN            NaN   

        5001|EIP [mbar]  5001|MVleak [L/min]  5001|RRspon [1/min]  \
0                   NaN                  NaN                  NaN   
1                  16.0                  0.0                  NaN   
2                   NaN                  NaN                  NaN   
3                  16.0                  0.0                  NaN   
4                   NaN                  NaN                  NaN   
...                 ...                  ...                  ...   
209895              NaN                  0.0                  NaN   
209896      

##### Now complete all the tasks we did last time

In [37]:
%%time

import datetime
import os
import pandas as pd

flist = [fle for fle in os.listdir('data') if 'slow_Measurement' in fle]

data_dict = {} # Creates an empty dictionary

for file in filelist:
    
    print(datetime.datetime.now(), file)
    path = os.path.join('data', file,)
    tag = file[11:-25]
    data_dict[tag] = pd.read_csv(path, parse_dates = [['Date', 'Time']])
    data_dict[tag] = data_dict[tag].set_index('Date_Time')
    
    new_columns = [item[5:] for item in data_dict[tag].columns if item.startswith('5001')]
    new_columns = ['Time [ms]', 'Rel.Time [s]'] + new_columns
    data_dict[tag].columns = new_columns
    
    data_dict[tag] = data_dict[tag].resample('1S').mean()
    
    columns_to_drop = ['Tispon [s]', 'I:Espon (I-Part) [no unit]', 'I:Espon (E-Part) [no unit]']
    data_dict[tag] = data_dict[tag].drop(columns_to_drop, axis = 1)

2021-04-20 09:12:23.887329 CsvLogBase_2019-01-14_124200.144_slow_Measurement.csv.zip
2021-04-20 09:12:25.132171 CsvLogBase_2019-01-16_090910.423_slow_Measurement.csv.zip
2021-04-20 09:12:26.248797 CsvLogBase_2020-11-02_134238.904_slow_Measurement.csv.zip
CPU times: user 4.97 s, sys: 1.12 s, total: 6.08 s
Wall time: 6.1 s


In [38]:
data_dict['2019-01-14_124200.144']

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2019-01-14 12:42:01  1.547470e+12           0.0         0.15         0.13   
2019-01-14 12:42:02  1.547470e+12           1.0         0.15         0.13   
2019-01-14 12:42:03  1.547470e+12           2.0         0.15         0.13   
2019-01-14 12:42:04  1.547470e+12           3.0         0.15         0.13   
2019-01-14 12:42:05  1.547470e+12           4.0         0.15         0.13   
...                           ...           ...          ...          ...   
2019-01-15 17:50:50  1.547575e+12      104929.0         0.00         2.38   
2019-01-15 17:50:51  1.547575e+12      104930.0         0.00         2.38   
2019-01-15 17:50:52  1.547575e+12      104931.0         0.00         2.38   
2019-01-15 17:50:53  1.547575e+12      104932.0         0.00         2.38   
2019-01-15 17:50:54  1.547575e+12      104933.0         0.00         2.38   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2019-01-14 12:42:01          0.29         142.0              0.0   
2019-01-14 12:42:02          0.29         152.0              0.0   
2019-01-14 12:42:03          0.29         152.0              0.0   
2019-01-14 12:42:04          0.27         152.0              0.0   
2019-01-14 12:42:05          0.27         143.0              0.0   
...                           ...           ...              ...   
2019-01-15 17:50:50           NaN           NaN              0.0   
2019-01-15 17:50:51           NaN           NaN              0.0   
2019-01-15 17:50:52           NaN           NaN              0.0   
2019-01-15 17:50:53           NaN           NaN              0.0   
2019-01-15 17:50:54           NaN           NaN              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2019-01-14 12:42:01            129.0             0.15              6.5  ...   
2019-01-14 12:42:02            139.0             0.15              6.5  ...   
2019-01-14 12:42:03            139.0             0.15              6.5  ...   
2019-01-14 12:42:04            139.0             0.15              6.5  ...   
2019-01-14 12:42:05            129.0             0.15              6.5  ...   
...                              ...              ...              ...  ...   
2019-01-15 17:50:50              NaN             0.00              NaN  ...   
2019-01-15 17:50:51              NaN             0.00              NaN  ...   
2019-01-15 17:50:52              NaN             0.00              NaN  ...   
2019-01-15 17:50:53              NaN             0.00              NaN  ...   
2019-01-15 17:50:54              NaN             0.00              NaN  ...   

                     VTspon [mL]  E [mbar/L]  TC [s]  TCe [s]  \
Date_Time                                                       
2019-01-14 12:42:01          0.0      3393.0    0.04     0.11   
2019-01-14 12:42:02          0.0      3393.0    0.04     0.11   
2019-01-14 12:42:03          0.0      3429.0    0.04     0.11   
2019-01-14 12:42:04          0.0      3754.0    0.04     0.11   
2019-01-14 12:42:05          0.0      3528.0    0.04     0.11   
...                          ...         ...     ...      ...   
2019-01-15 17:50:50          0.0         NaN     NaN      NaN   
2019-01-15 17:50:51          0.0         NaN     NaN      NaN   
2019-01-15 17:50:52          0.0         NaN     NaN      NaN   
2019-01-15 17:50:53          0.0         NaN     NaN      NaN   
2019-01-15 17:50:54          0.0         NaN     NaN      NaN   

                     C20/Cdyn [no unit]  VTe [mL]  VTi [mL]  EIP [mbar]  \
Date_Time                                                                 
2019-01-14 12:42:01                0.58       2.7       2.4        16.0   
2019-01-14 12:42:02                0.6

In [39]:
[key for key in data_dict.keys()]

['2019-01-14_124200.144', '2019-01-16_090910.423', '2020-11-02_134238.904']

In [40]:
[len(value) for value in data_dict.values()]

[104934, 101443, 344677]

In [41]:
[value.shape for value in data_dict.values()]

[(104934, 42), (101443, 42), (344677, 42)]

#### Some parameters (`VTs, MVs, Cdyn`) only make sense if normalised to body weight

In [42]:
data_dict['2019-01-14_124200.144'].columns

Index(['Time [ms]', 'Rel.Time [s]', 'MVe [L/min]', 'MVi [L/min]',
       'Cdyn [L/bar]', 'R [mbar/L/s]', 'MVespon [L/min]', 'Rpat [mbar/L/s]',
       'MVemand [L/min]', 'FlowDev [L/min]', 'VTmand [mL]', 'r2 [no unit]',
       'VTispon [mL]', 'Pmin [mbar]', 'Pmean [mbar]', 'PEEP [mbar]',
       'RRmand [1/min]', 'PIP [mbar]', 'VTmand [L]', 'VTspon [L]',
       'VTemand [mL]', 'VTespon [mL]', 'VTimand [mL]', 'VT [mL]', '% leak [%]',
       '% MVspon [%]', 'MV [L/min]', 'RRtrig [1/min]', 'RR [1/min]',
       'I (I:E) [no unit]', 'E (I:E) [no unit]', 'FiO2 [%]', 'VTspon [mL]',
       'E [mbar/L]', 'TC [s]', 'TCe [s]', 'C20/Cdyn [no unit]', 'VTe [mL]',
       'VTi [mL]', 'EIP [mbar]', 'MVleak [L/min]', 'RRspon [1/min]'],
      dtype='object')

In [43]:
to_normalise = ['MVe [L/min]', 'MVi [L/min]', 'Cdyn [L/bar]', 'MVespon [L/min]', 'MVemand [L/min]', 'VTmand [mL]', 
    'VTispon [mL]', 'VTmand [L]', 'VTspon [L]', 'VTemand [mL]', 'VTespon [mL]', 'VTimand [mL]', 'VT [mL]', 
    'MV [L/min]',  'VTspon [mL]', 'VTe [mL]', 'VTi [mL]', 'MVleak [L/min]',]

In [44]:
data_dict.keys()

dict_keys(['2019-01-14_124200.144', '2019-01-16_090910.423', '2020-11-02_134238.904'])

In [45]:
# Weights in kilogram
weights = {'2019-01-14_124200.144' : 0.575, '2019-01-16_090910.423' : 0.575 , '2020-11-02_134238.904' : 775}

In [46]:
for recording in data_dict: # I could have written datadict.keys(), it is the same
    for par in to_normalise:
        data_dict[recording][f'{par[:-1]}/kg{par[-1]}'] = data_dict[recording][par] / weights[recording]

In [47]:
data_dict['2019-01-14_124200.144']

Time [ms]  Rel.Time [s]  MVe [L/min]  MVi [L/min]  \
Date_Time                                                                   
2019-01-14 12:42:01  1.547470e+12           0.0         0.15         0.13   
2019-01-14 12:42:02  1.547470e+12           1.0         0.15         0.13   
2019-01-14 12:42:03  1.547470e+12           2.0         0.15         0.13   
2019-01-14 12:42:04  1.547470e+12           3.0         0.15         0.13   
2019-01-14 12:42:05  1.547470e+12           4.0         0.15         0.13   
...                           ...           ...          ...          ...   
2019-01-15 17:50:50  1.547575e+12      104929.0         0.00         2.38   
2019-01-15 17:50:51  1.547575e+12      104930.0         0.00         2.38   
2019-01-15 17:50:52  1.547575e+12      104931.0         0.00         2.38   
2019-01-15 17:50:53  1.547575e+12      104932.0         0.00         2.38   
2019-01-15 17:50:54  1.547575e+12      104933.0         0.00         2.38   

                     Cdyn [L/bar]  R [mbar/L/s]  MVespon [L/min]  \
Date_Time                                                          
2019-01-14 12:42:01          0.29         142.0              0.0   
2019-01-14 12:42:02          0.29         152.0              0.0   
2019-01-14 12:42:03          0.29         152.0              0.0   
2019-01-14 12:42:04          0.27         152.0              0.0   
2019-01-14 12:42:05          0.27         143.0              0.0   
...                           ...           ...              ...   
2019-01-15 17:50:50           NaN           NaN              0.0   
2019-01-15 17:50:51           NaN           NaN              0.0   
2019-01-15 17:50:52           NaN           NaN              0.0   
2019-01-15 17:50:53           NaN           NaN              0.0   
2019-01-15 17:50:54           NaN           NaN              0.0   

                     Rpat [mbar/L/s]  MVemand [L/min]  FlowDev [L/min]  ...  \
Date_Time                                                               ...   
2019-01-14 12:42:01            129.0             0.15              6.5  ...   
2019-01-14 12:42:02            139.0             0.15              6.5  ...   
2019-01-14 12:42:03            139.0             0.15              6.5  ...   
2019-01-14 12:42:04            139.0             0.15              6.5  ...   
2019-01-14 12:42:05            129.0             0.15              6.5  ...   
...                              ...              ...              ...  ...   
2019-01-15 17:50:50              NaN             0.00              NaN  ...   
2019-01-15 17:50:51              NaN             0.00              NaN  ...   
2019-01-15 17:50:52              NaN             0.00              NaN  ...   
2019-01-15 17:50:53              NaN             0.00              NaN  ...   
2019-01-15 17:50:54              NaN             0.00              NaN  ...   

                     VTspon [L/kg]  VTemand [mL/kg]  VTespon [mL/kg]  \
Date_Time                                                              
2019-01-14 12:42:01            0.0         4.695652              0.0   
2019-01-14 12:42:02            0.0         4.869565              0.0   
2019-01-14 12:42:03            0.0         4.869565              0.0   
2019-01-14 12:42:04            0.0         5.043478              0.0   
2019-01-14 12:42:05            0.0         5.043478              0.0   
...                            ...              ...              ...   
2019-01-15 17:50:50            0.0              NaN              0.0   
2019-01-15 17:50:51            0.0              NaN              0.0   
2019-01-15 17:50:52            0.0              NaN              0.0   
2019-01-15 17:50:53            0.0              NaN              0.0   
2019-01-15 17:50:54            0.0              NaN              0.0   

                     VTimand [mL/kg]  VT [mL/kg]  MV [L/min/kg]  \
Date_Time                                                         
2019-01-14 12:42:01         4.173913  

#### Export them as serialised binary data  - `pickle`

In [48]:
%%time

import pickle

for key in data_dict.keys():
    print(key)
    path = os.path.join('results', f'data_processed_{key}.pickle')
    filehandle = open(path, 'wb')
    pickle.dump(data_dict[key], filehandle)
    filehandle.close()

2019-01-14_124200.144
2019-01-16_090910.423
2020-11-02_134238.904
CPU times: user 85.9 ms, sys: 170 ms, total: 255 ms
Wall time: 343 ms


### To be continued... 